### PR TITLE
[Draft] Add support for WCH CH32V SoC/CH32V307V-EVB-R1 board

### PR DESCRIPTION
**Pull request by soburi**
Add support for WCH CH32V307, and CH32V307V-EVB-R1 which is the evaluation board of this SoC.
Currently, Blinky and Hello World sample worked and it be able to debug and flash.

CH32V307 is built with a RISC-V core and STM32 cloned peripherals.
It is a similar configuration to [GigaDevice GD32VF103](https://docs.zephyrproject.org/3.1.0/boards/riscv/gd32vf103v_eval/doc/index.html).

## HAL problems

I think it can use the gd32 driver by adding a modification to vendor-supplied HAL.
(Add APIs that are compatible with GD32 HAL's one.)
Or maybe good to create a "generic STM32 clone HAL".

This point will need to discuss.

## OpenOCD problems

The link is OpenOCD binaries that support CH32V307.
http://file.mounriver.com/tools/MRS_Toolchain_Linux_x64_V1.50.tar.xz

The source is not officially provided.
It will probably also become a problem.
(unofficially https://github.com/kprasadvnsi/riscv-openocd-wch)

So, I think complete this PR more a little time.

### CODE REVIEW

File: arch/riscv/Kconfig
Comment (carlocaione) [inline arch/riscv/Kconfig:126]: I'm fine with adding this but adding `weak` symbols and blurring the distinction between PLIC and CLIC is something that must be justified a bit better. Why do you need this? To save some code or do you actually need this to be able to implement support for this board?

File: west.yml
Comment (carlocaione) [inline west.yml:25]: we usually do not pull directly from outside repos. See the other modules in this file.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: west.yml
Comment (soburi) [reply west.yml:25]: I need to submit it to zephyr as the new module.
I will update it after approve it.

File: arch/riscv/Kconfig
Comment (soburi) [reply arch/riscv/Kconfig]: I think it make improvement to introduce the weak symbol, but that is not mainly intention in this PR.
Rewrite it based on current code.

I consider to create another PR to introduce the weak symbol.

**Review by gmarull — CHANGES_REQUESTED** [review]
Definitely a no-go to re-use GD32 drivers. While CH devices are yet another ST clone, we can't re-use Zephyr drivers because we're using HALs underneath.

It's interesting because we could probably have generic drivers for ST/GD/CH if they were native, but definitely not with current approach. But still, these clones are not 100% equal, registers are named differently to start with...

**Comment by tagunil** [comment]
> Definitely a no-go to re-use GD32 drivers. While CH devices are yet another ST clone, we can't re-use Zephyr drivers because we're using HALs underneath.
> 
> It's interesting because we could probably have generic drivers for ST/GD/CH if they were native, but definitely not with current approach. But still, these clones are not 100% equal, registers are named differently to start with...

Could you please point me to any specific differences or other issues that could prevent using ST peripheral drivers for these clones?

**Comment by gmarull** [comment]
> > Definitely a no-go to re-use GD32 drivers. While CH devices are yet another ST clone, we can't re-use Zephyr drivers because we're using HALs underneath.
> > It's interesting because we could probably have generic drivers for ST/GD/CH if they were native, but definitely not with current approach. But still, these clones are not 100% equal, registers are named differently to start with...
> 
> Could you please point me to any specific differences or other issues that could prevent using ST peripheral drivers for these clones?

HAL license?

**Comment by tagunil** [comment]
> > > Definitely a no-go to re-use GD32 drivers. While CH devices are yet another ST clone, we can't re-use Zephyr drivers because we're using HALs underneath.
> > > It's interesting because we could probably have generic drivers for ST/GD/CH if they were native, but definitely not with current approach. But still, these clones are not 100% equal, registers are named differently to start with...
> > 
> > Could you please point me to any specific differences or other issues that could prevent using ST peripheral drivers for these clones?
> 
> HAL license?

As far as I can see, HALs are licensed under 3-clause BSD.

**Comment by gmarull** [comment]
> > > > Definitely a no-go to re-use GD32 drivers. While CH devices are yet another ST clone, we can't re-use Zephyr drivers because we're using HALs underneath.
> > > > It's interesting because we could probably have generic drivers for ST/GD/CH if they were native, but definitely not with current approach. But still, these clones are not 100% equal, registers are named differently to start with...
> > > 
> > > 
> > > Could you please point me to any specific differences or other issues that could prevent using ST peripheral drivers for these clones?
> > 
> > 
> > HAL license?
> 
> As far as I can see, HALs are licensed under 3-clause BSD.

Right, so in theory it could be done. However, in case of GD32 for example, register names are different, and part numbers do not follow STM32, see for example the big IP differences between F403/F450. I'd only go for unified drivers if they are implemented natively, where you have freedom on how to organize code properly and account for multiple vendors differences. I doubt that ST maintainers have any interest in going that path, even though it would be nice to see.

**Comment by tagunil** [comment]
> > > > > Definitely a no-go to re-use GD32 drivers. While CH devices are yet another ST clone, we can't re-use Zephyr drivers because we're using HALs underneath.
> > > > > It's interesting because we could probably have generic drivers for ST/GD/CH if they were native, but definitely not with current approach. But still, these clones are not 100% equal, registers are named differently to start with...
> > > > 
> > > > 
> > > > Could you please point me to any specific differences or other issues that could prevent using ST peripheral drivers for these clones?
> > > 
> > > 
> > > HAL license?
> > 
> > 
> > As far as I can see, HALs are licensed under 3-clause BSD.
> 
> Right, so in theory it could be done. However, in case of GD32 for example, register names are different, and part numbers do not follow STM32, see for example the big IP differences between F403/F450. I'd only go for unified drivers if they are implemented natively, where you have freedom on how to organize code properly and account for multiple vendors differences. I doubt that ST maintainers have any interest in going that path, even though it would be nice to see.

I think these differences could be accounted for even with current approach, at least for simple peripherals like UARTs and timers. But I agree that this path is infeasible without some degree of maintainer support.

**Comment by soburi** [comment]
@tagunil 

Ideally, libopencm3 is a generic firmware for STM32clone, so I'd like to use it, but unfortunately the license is GPL, so I can't incorporate it into zephyr.

Since the low-level HAL of STM32 is BSD licensed, It may be possible to support combined with it and 'forked' gd32 driver.
However, even in this case, it seems that various F103 clones that are highly compatible with the original can be used, but I think that clones such as F4XX will require individual support. (Otherwise we'll be in #ifdef hell!)

**Comment by findlayfeng** [comment]
@soburi Hello, is this PR still working?
I can do something.

**Comment by soburi** [comment]
@findlayfeng 

As an issue, I'm worried about how better way to make HAL, but I can't find a good approach, so I've stuck.

Of course, I can solve it by simply importing the HAL provided by WCH, but I'm consider if I can create a common HAL for STM32(F103) clones.

**Comment by findlayfeng** [comment]
I think the HAL provided by WCH should be imported directly.
although it looks very similar to stm32(f103), we are not maintainers of the hardware and don't need to consider whether they are consistent.

**Pull request closed without merging**
